src/nix/installables.cc

```diff
@@ -440,6 +440,16 @@ ref<eval_cache::EvalCache> openEvalCache(
         }));
 }
 
+static std::string showAttrPaths(const std::vector<std::string> & paths)
+{
+    std::string s;
```

Comment:
This might be a bit nit-picky but I think we can improve readability and reduce duplicate code (e.g. the concat and quote code) by just implementing it like this:

```cpp
static std::string showAttrPaths(const std::vector<std::string> & paths)
{
        std::string s;
        auto quotedPaths = quoteStrings(paths);
        auto it = quotedPaths.begin();

        if (it != quotedPaths.end())
                s = *it++;

        if (it != quotedPaths.end()) {
                auto rem = concatStringsSep(", ", std::vector<std::string>(it, quotedPaths.end()));
                s += " or " + rem;
        }

        return s;
}
```

This removes the ternary operator and the line that contains three different statements that mostly consists of quote & operator characters. Worst case this will create a copy of the remaining elements of the vector but IMHO that shouldn't hurt given the benefits.

Reply:
I sort of got nerd-sniped reading this and tried to write something that i can parse quite easily:

```c++
/**
 * Takes a list of paths and turns them into a comma separated list
 * of the format shown below:
 *
 * { }               -> ""
 * { "x" }           -> "'x'"
 * { "x", "y" }      -> "'x' or 'y'
 * { "x", "y", "z" } -> "'x' or 'y', 'z'"
 */
static std::string showAttrPaths(const std::vector<std::string> & paths)
{
    if (paths.size() == 0)
        return "";

    if (paths.size() == 1)
        return "'" + paths[0] + "'";

    auto quotedPaths = quoteStrings(paths);
    auto it = std::next(quotedPaths.begin());

    return "'"
        + paths[0]
        + "' or "
        + concatStringsSep(", ", std::vector<std::string>(it, quotedPaths.end()));
}
```

By the way: In the case of an empty `paths` vector, should the result really just be an empty string?